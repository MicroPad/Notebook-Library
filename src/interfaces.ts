export interface Parent {
	title: string;
}
