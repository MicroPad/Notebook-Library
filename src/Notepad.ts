import { format, parse } from 'date-fns';
import { Builder } from 'xml2js';
import FlatNotepad, { FlatSection } from './FlatNotepad';
import Note, { MarkdownNote } from './Note';
import { NotepadShell } from "./interfaces";
import { encrypt, EncryptionMethod } from './crypto';
import { LAST_MODIFIED_FORMAT } from './date-formats';
import Asset from './Asset';
import Section from './Section';

export type NotepadOptions = {
	lastModified?: Date;
	sections?: Section[];
	notepadAssets?: string[];
	assets?: Asset[];
	crypto?: EncryptionMethod;
};

/**
 * This class is identical in structure to the old Notepad class from the original parser.
 * This represents the notepad as a tree. If you're looking for a flatter structure you can use {@link FlatNotepad}.
 *
 * Something to remember is that all operations on this class like addSection, will return a <strong>new</strong>
 * object of this class, and not modify the existing one.
 */
export default class Notepad implements NotepadShell {
	public readonly lastModified: string;
	public readonly sections: Section[];
	public readonly notepadAssets: string[];
	public readonly assets: Asset[];
	public readonly crypto?: EncryptionMethod;

	constructor(
		public readonly title: string,
		opts: NotepadOptions = {}
	) {
		this.lastModified = format(opts.lastModified || new Date(), LAST_MODIFIED_FORMAT);
		this.sections = opts.sections || [];
		this.notepadAssets = opts.notepadAssets || [];
		this.assets = opts.assets || [];
		if (opts.crypto) this.crypto = opts.crypto;
	}

	public addSection(section: Section): Notepad {
		const notepad = this.clone({
			sections: [
				...this.sections,
				section
			]
		});
		section.parent = notepad;

		return notepad;
	}

	public addAsset(asset: Asset): Notepad {
		return this.clone({
			assets: [
				...this.assets,
				asset
			],
			notepadAssets: [
				...this.notepadAssets,
				asset.uuid
			]
		});
	}

	/**
	 * This updates the lastModified date on the notepad. This date is used for syncing so it's important
	 * to call this method whenever a change is made that will need to be synced.
	 * @param {Date} lastModified
	 * @returns {Notepad}
	 */
	public modified(lastModified: Date = new Date()): Notepad {
		return this.clone({
			lastModified
		});
	}

	/**
	 * @param {string} query Can either be a title-search or a hashtag-search
	 * @returns {Note[]}
	 */
	public search(query: string): Note[] {
		return this.sections
			.map(s => s.search(query))
			.reduce((acc, val) => acc.concat(val), []);
	}

	public async toJson(passkey?: string): Promise<string> {
		let notepad: Notepad = {
			...this,
			assets: undefined,

			// If we're given a passkey but encryption hasn't been setup on the notepad, set it up
			crypto: (!this.crypto && !!passkey) ? 'AES-256-GZ' : this.crypto
		};

		let notepadToStringify: NotepadShell = notepad;
		if (!!notepad.crypto && !!passkey) notepadToStringify = await encrypt(notepad, passkey);

		if (!!notepad.crypto && !passkey) {
			notepadToStringify = { ...notepadToStringify, crypto: undefined };
		}

		return JSON.stringify(notepadToStringify, (key, value) => {
			return (key === 'parent') ? undefined : value;
		});
	}

	public async toXml(): Promise<string> {
		const builder = new Builder({
			cdata: true,
			renderOpts: {
				'pretty': false
			},
			xmldec: {
				version: '1.0',
				encoding: 'UTF-8',
				standalone: false
			}
		});

		// Generate the XML
		const obj = await this.toXmlObject();
		return builder.buildObject(obj).replace(/&#xD;/g, '');
	}

	public flatten(): FlatNotepad {
		let notepad = new FlatNotepad(this.title, {
			lastModified: parse(this.lastModified, LAST_MODIFIED_FORMAT, new Date()),
			notepadAssets: this.notepadAssets,
			crypto: this.crypto
		});

		const flattenSection = (section: Section) => {
			let flat: FlatSection = { title: section.title, internalRef: section.internalRef };
			if (section.parent) flat.parentRef = (section.parent as Section).internalRef;

			// Add this flat section
			notepad = notepad.addSection(flat);
			section.notes.forEach(n => notepad = notepad.addNote(n));

			// Add all of its children recursively
			section.sections.forEach(s => flattenSection(s));
		};
		this.sections.forEach(s => flattenSection(s));

		return notepad;
	}

	public async toMarkdown(assets: Asset[]): Promise<MarkdownNote[]> {
		return (
			await Promise.all(
				this.sections.map(s => s.toMarkdown(assets))
			)
		).reduce((acc, val) => acc.concat(val), []);
	}

	public clone(opts: Partial<NotepadOptions> = {}, title: string = this.title): Notepad {
		return new Notepad(title, {
			lastModified: parse(this.lastModified, LAST_MODIFIED_FORMAT, new Date()),
			sections: [...this.sections],
			notepadAssets: [...this.notepadAssets],
			assets: [...this.assets],
			crypto: this.crypto,
			...opts
		});
	}

	private async toXmlObject(): Promise<object> {
		return {
			notepad: {
				$: {
					'xsi:noNamespaceSchemaLocation': 'https://getmicropad.com/schema.xsd',
					title: this.title,
					lastModified: this.lastModified,
					'xmlns:xsi': 'http://www.w3.org/2001/XMLSchema-instance'
				},
				assets: [
					{
						asset: await Promise.all(this.assets.map(a => a.toXmlObject()))
					}
				],
				section: this.sections.map(s => s.toXmlObject().section)
			}
		};
	}
}
