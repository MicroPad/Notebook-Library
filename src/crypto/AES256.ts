import { EncryptionMethodImpl } from './index';
import { NotepadShell } from '../interfaces';
import Notepad from '../Notepad';
import scrypt from 'scrypt-js';
import buffer from 'scrypt-js/thirdparty/buffer';
import * as AES from 'aes-js';
import { Translators } from '../Translators';
import stringify from 'json-stringify-safe';


export class AES256 implements EncryptionMethodImpl {
	async decrypt(notepad: NotepadShell, passkey: string): Promise<Notepad> {
		const cipherText = notepad.sections as string;
		const key = await this.keyGenerator(passkey);
		const controller = new AES.ModeOfOperation.ctr(key);

		const plainText = AES.utils.utf8.fromBytes(controller.decrypt(AES.utils.hex.toBytes(cipherText)));
		notepad = { ...notepad, sections: JSON.parse(plainText) };

		return Translators.Json.toNotepadFromNotepad(notepad);
	}

	async encrypt(notepad: Notepad, passkey: string): Promise<NotepadShell> {
		const plainText = stringify(notepad.sections);
		const key = await this.keyGenerator(passkey);
		const controller = new AES.ModeOfOperation.ctr(key);

		const cipherText = AES.utils.hex.fromBytes(controller.encrypt(AES.utils.utf8.toBytes(plainText)));
		return { ...notepad, sections: cipherText };
	}

	private keyGenerator(passkey: string): Promise<ReadonlyArray<number>> {
		return new Promise<ReadonlyArray<number>>((resolve, reject) => {
			passkey = passkey.normalize('NFKC');
			const passkeyBuff = new buffer.SlowBuffer(passkey);

			scrypt(passkeyBuff, new buffer.SlowBuffer(''), 1024, 8, 1, 32, (err, progress, key) => {
				if (!!err) reject(err);
				if (!!key) resolve(key);
			});
		});
	}
}