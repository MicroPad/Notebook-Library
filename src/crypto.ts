import { NotepadShell } from './interfaces';
import Notepad from './Notepad';
import { AES256 } from './AES256';
import { AES256GZ } from './AES256GZ';

export type EncryptionMethod = 'AES-256' | 'AES-256-GZ';

export async function decrypt(notepad: NotepadShell, passkey: string): Promise<Notepad> {
	return getMethod(notepad).decrypt(notepad, passkey);
}

export async function encrypt(notepad: Notepad, passkey: string): Promise<NotepadShell> {
	return getMethod(notepad).encrypt(notepad, passkey);
}

function getMethod(notepad: NotepadShell): EncryptionMethodImpl {
	if (!notepad.crypto) throw new Error(`This notepad isn't encrypted.`);

	const method = methods[notepad.crypto];
	if (!method) throw new Error(`No such method: '${notepad.crypto}' exists.`);

	return method;
}

const methods: { [K in EncryptionMethod]: EncryptionMethodImpl } = {
	'AES-256': new AES256(),
	'AES-256-GZ': new AES256GZ()
};

export interface EncryptionMethodImpl {
	encrypt(notepad: Notepad, passkey: string): Promise<NotepadShell>;
	decrypt(notepad: NotepadShell, passkey: string): Promise<Notepad>;
}
