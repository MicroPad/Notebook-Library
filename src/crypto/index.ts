import { NotepadShell } from '../interfaces';
import Notepad from '../Notepad';
import { AES256 } from './AES256';

export type EncryptionMethod = 'AES-256';

export async function decrypt(notepad: NotepadShell, passkey: string): Promise<Notepad> {
	return getMethod(notepad).decrypt(notepad, passkey);
}

export async function encrypt(notepad: Notepad, passkey: string): Promise<NotepadShell> {
	return getMethod(notepad).encrypt(notepad, passkey);
}

function getMethod(notepad: NotepadShell): EncryptionMethodImpl {
	if (!notepad.crypto) throw new Error(`This notepad isn't encrypted.`);

	const method = methods[notepad.crypto];
	if (!method) throw new Error(`No such method: '${notepad.crypto}' exists.`);

	return method;
}

const methods: { [K in EncryptionMethod]: EncryptionMethodImpl } = {
	'AES-256': new AES256()
};

export interface EncryptionMethodImpl {
	encrypt(notepad: Notepad, passkey: string): Promise<NotepadShell>;
	decrypt(notepad: NotepadShell, passkey: string): Promise<Notepad>;
}
