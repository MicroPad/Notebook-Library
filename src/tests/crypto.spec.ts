import { decrypt, encrypt } from '../crypto';
import Notepad, { NotepadOptions } from '../Notepad';
import { NotepadShell } from '../interfaces';
import { TestUtils } from './TestUtils';
import { Section } from '../index';

describe('Crypto', () => {
	let notepad: Notepad;
	let encryptedNotepad: NotepadShell;
	const passkey = 'this is a bad password';

	beforeEach(async () => {
		notepad = new Notepad('test', getOptions());
		notepad = notepad.clone({
			sections: [
				notepad.sections[0].addNote(TestUtils.makeNote('test note'))
			],
			crypto: 'AES-256'
		});

		encryptedNotepad = { ...notepad, sections: 'a4bc7633fe1c3ab0af9b3735656ec452c359d65cc7f372812d49e375e1fd7583bd181b70701bffa44fd747b1b5097f4a41958c54330aa9ca0f186cc626afd8dd352a95f8f1e4945d51ca79447334ea132f482cf3da8ae1e10851bac53bb3842e99ad5039455b6536ba53be619a6cf5ba73681a39da6041682318eef298df5ec824a2be858a6c96d2109e7e7344a19f4e0e0ba37b49f71a25b0fe9b3e588565099d8e400f4d737a822368f3c652d9e67484fb68' };
	});

	describe('encrypt', () => {
		it('AES256', async () => {
			// Arrange
			// Act
			const res = await encrypt(notepad, passkey);

			// Assert
			expect(res).toMatchSnapshot();
		});
	});

	describe('decrypt', () => {
		it('AES256', async () => {
			// Arrange
			// Act
			const res = await decrypt(encryptedNotepad, passkey);

			// Assert
			expect(res).toMatchSnapshot();
		});
	});
});

function getOptions(): NotepadOptions {
	const testSection = new Section('test section');
	(<any> testSection).internalRef = 'abc';

	return {
		lastModified: new Date(1),
		sections: [testSection],

		assets: [TestUtils.makeAsset()],
		notepadAssets: ['test']
	};
}
