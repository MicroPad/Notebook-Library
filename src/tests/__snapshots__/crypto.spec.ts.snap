// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Crypto decrypt AES256 1`] = `
Notepad {
  "assets": Array [],
  "lastModified": "1970-01-01T12:00:00.001+12:00",
  "notepadAssets": Array [
    "test",
  ],
  "sections": Array [
    Section {
      "internalRef": "abc",
      "notes": Array [
        Note {
          "bibliography": Array [],
          "elements": Array [],
          "internalRef": "abc",
          "parent": [Circular],
          "time": 1,
          "title": "test note",
        },
      ],
      "parent": [Circular],
      "sections": Array [],
      "title": "test section",
    },
  ],
  "title": "test",
}
`;

exports[`Crypto encrypt AES256 1`] = `
Object {
  "assets": Array [
    Asset {
      "data": Blob {
        Symbol(impl): BlobImpl {
          "_buffer": Object {
            "data": Array [
              116,
              101,
              115,
              116,
            ],
            "type": "Buffer",
          },
          "type": "",
          Symbol(wrapper): [Circular],
        },
      },
      "uuid": "abc",
    },
  ],
  "crypto": "AES-256",
  "lastModified": "1970-01-01T12:00:00.001+12:00",
  "notepadAssets": Array [
    "test",
  ],
  "sections": "a4bc7633fe1c3ab0af9b3735656ec452c359d65cc7f372812d49e375e1fd7583bd181b70701bffa44fd747b1b5097f4a41958c54330aa9ca0f186cc626afd8dd352a95f8f1e4945d51ca79447334ea132f482cf3da8ae1e10851bac53bb3842e99ad5039455b6536ba53be619a6cf5ba73681a39da6041682318eef298df5ec824a2be858a6c96d2109e7e7344a19f4e0e0ba37b49f71a25b0fe9b3e588565099d8e400f4d737a822368f3c652d9e67484fb68",
  "title": "test",
}
`;
